subprojects/dependency-management/src/main/java/org/gradle/internal/component/external/model/AbstractLazyModuleComponentResolveMetadata.java

```diff
@@ -87,21 +93,71 @@ public VariantMetadataRules getVariantMetadataRules() {
+        ImmutableList<? extends ComponentVariant> variants = getVariants();
         if (variants.isEmpty()) {
-            return maybeDeriveVariants();
+            return addVariantsByRule(maybeDeriveVariants());
         }
-        ImmutableList.Builder<ConfigurationMetadata> configurations = new ImmutableList.Builder<ConfigurationMetadata>();
+        ImmutableList.Builder<ConfigurationMetadata> configurations = new ImmutableList.Builder<>();
         for (ComponentVariant variant : variants) {
             configurations.add(new LazyVariantBackedConfigurationMetadata(getId(), variant, getAttributes(), getAttributesFactory(), variantMetadataRules));
         }
-        return Optional.<ImmutableList<? extends ConfigurationMetadata>>of(configurations.build());
+        return addVariantsByRule(Optional.of(configurations.build()));
+
+    }
+
+    private Optional<ImmutableList<? extends ConfigurationMetadata>> addVariantsByRule(Optional<ImmutableList<? extends ConfigurationMetadata>> variants) {
+        if (variantMetadataRules.getAdditionalVariants().isEmpty()) {
+            return variants;
+        }
+        Map<String, ConfigurationMetadata> variantsByName = variants.or(ImmutableList.of()).stream().collect(Collectors.toMap(ConfigurationMetadata::getName, Function.identity()));
+        ImmutableList.Builder<ConfigurationMetadata> builder = new ImmutableList.Builder<>();
+        if (variants.isPresent()) {
+            builder.addAll(variants.get());
+        }
+        for (Map.Entry<String, String> variantName : variantMetadataRules.getAdditionalVariants().entrySet()) {
+            String baseName = variantName.getValue();
+            ConfigurationMetadata base = null;
+            if (baseName != null) {
+                if (variants.isPresent()) {
+                    base = variantsByName.get(baseName);
+                    if (base == null) {
+                        throw new InvalidUserDataException("Variant '" + baseName + "' not defined in module " + getId().getDisplayName());
+                    }
+                } else {
+                    base = getConfiguration(baseName);
+                    if (base == null) {
+                        throw new InvalidUserDataException("Configuration '" + baseName + "' not defined in module " + getId().getDisplayName());
+                    }
+                }
+            }
+            ImmutableAttributes attributes;
+            ImmutableCapabilities capabilities;
+            List<? extends ModuleDependencyMetadata> dependencies;
+            ImmutableList<? extends ComponentArtifactMetadata> artifacts;
+            if (base instanceof ModuleConfigurationMetadata) {
+                attributes = base.getAttributes();
+                capabilities = (ImmutableCapabilities) base.getCapabilities();
+                dependencies = ((ModuleConfigurationMetadata) base).getDependencies();
+                artifacts = base.getArtifacts();
```

Comment:
I Just realized that this code block is too eager as it will execute all rules that match the _base variant_. Instead of calling the getters here, the base variant should be wrapped and delegated to if the new variant is actually used.